test/testlib/minikube_utilities.go

```diff
@@ -29,21 +30,59 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+/** Lists of the teardown and diagnostic teardown funcs */
 var teardownLists = make(map[string][]func())
+var diagnosticTeardownLists = make(map[string][]func())
 
 /**
- * add a teardown function to the named list - for later execution
+ * add a teardown function to the named list - for deferred execution.
+ *
+ * The teardown functions are called in reverse order of insertion, by a call to Teardown(name).
+ *
+ * The typical idiom is:
+ * <pre>
+ *   testlib.AddTeardown("DATABASE", func() { ...})
+ *   // possibly more testlib.AddTeardown("DATABASE", func() { ... })
+ *   defer testlib.Teardown("DATABASE")
+ * <pre>
  */
 func AddTeardown(name string, teardownFunc func()) {
 	teardownLists[name] = append(teardownLists[name], teardownFunc)
 }
 
+/**
+ * add a diagnostic teardown func to be called before any other teardowns in the named list - to aid diagnostics/debugging.
+ * This allows a diagnostic teardown to do such things as:
+ * <ul>
+ *   <li>Generate logging and debug information immediately prior to resurce teardown
+ *   <li>call time.Sleep() to allow inspection and/or debugging of the exit state before teardown.
+ * <ul>
+ *
+ * NOTE: it is generally undesirable to add multiple diagnostic teardowns that sleep - so it would usually be best to
+ * add any Sleep() debug teardown to the innermost teardown list.
+ * Nonetheless, there are use-cases where multiple Sleep() teardowns are useful - to allow inspecting different
+ * intermediate states.
+ */
+func AddDiagnosticTeardown(name string, teardownFunc func()) {
+	diagnosticTeardownLists[name] = append(diagnosticTeardownLists[name], teardownFunc)
+}
+
 /**
  * Call the stored teardown functions in the named list, in the correct order (last-in-first-out)
+ *
+ * NOTE: Any DIAGNOSTIC teardowns - those added with AddDiagnosticTeardown() for this name - are called BEFORE any other teardowns for this name.
+ *
+ * The typical use of Teardown is with a deferred call:
+ * defer testlib.Teardown("SOME NAME")
+ * See: testlib.AddTeardown(); testlib.AddDiagnosticTeardown()
  */
 func Teardown(name string) {
+	// ensure both list and diagnostic list are removed.
+	defer func() { delete(diagnosticTeardownLists, name) }()
+	defer func() { delete(teardownLists, name) }()
+
 	list := teardownLists[name]
-	delete(teardownLists, name)
+	list = append(list, diagnosticTeardownLists[name]...) // append any diagnostic funcs - so they are called FIRST
```

Comment:
I would only execute diagnostic teardown if the test has failed. That way you can just add them and they will only print something if they are actually required. You would have to use `if t.hasFailed()`

Reply:
Is that a request to change AddDiagnosticTeardown to take a *testing.T argument, and check it before calling the diagnostic teardowns?

I have actually used diagnostic teardowns to dump the exit state to log to understand the sequence of events in situations where pods are restarted - so I would like making the t.Failed() test to be optional.

So I would be happy to take a *testing.T argument, but skip the test if the passed in *testing.T is nil.

Reply:
you could override it with an environmental variable? Something like `if t.Failed() || ALWAYS_PRINT_DIAGNOSE`

Reply:
In the expectation that we would normally only set that variable when running tests locally - and it would always take the if t.Failed() path in Travis?

That sounds fair...

Reply:
But as far as I can see, I still need to pass the *testing.T var as a parameter, since the *testing.T is not in the closure of the AddDiagnoseTeardown() function.

Reply:
Ok, I've added an EnvVar ALWAYS_RUN_DIAGNOSTIC_TEARDOWNS; plus a condition param to AddDiagnosticTeardown() to control whether diagnostic teardowns are run or not.

The condition param is a bit more flexible than just passing a *testing.T - but that also works.